### PR TITLE
[ads] Fixes search result ad landing page metrics are not reported on iOS

### PR DESCRIPTION
**Pull request by aseren**
Resolves https://github.com/brave/brave-browser/issues/41722

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Test case 1
- Join Brave Rewards
- Open search.brave.software
- Search for giraffe
- Tap the [STAGING] Brave Ads search result ad
   EXPECTATION: Search result ad landing page metrics are reported. There should be a log:
      `[ads] Maybe land on page for https://brave.com/brave-ads/?no_key_param= in 5 s`
      `[ads] Landed on page for https://brave.com/brave-ads/?no_key_param= on tab id %%%`

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift [hunk 1]
```diff
@@ -997,6 +997,15 @@ extension BrowserViewController: WKNavigationDelegate {
 
     rewards.reportTabNavigation(tabId: tab.rewardsId)
 
+    // Notify of tab changes after navigation completes but before notifying that
+    // the tab has loaded, so that any listeners can process the tab changes
+    // before the tab is considered loaded.
+    rewards.maybeNotifyTabDidChange(
```
Comment (aseren) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift:1003]: Moved `NotifyTabDidChange` and `NotifyTabDidLoad` notifications to `webView(_:didCommit:)` to align with the Desktop implementation.

File: ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift [hunk 2]
```diff
@@ -220,6 +220,46 @@ public class BraveRewards: PreferencesObserver {
     }
   }
 
+  // MARK: - Brave Ads Notifications
```
Comment (tmancey) [inline ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift:223]: This // MARK is in the wrong place, as it is related to push notification ads

File: ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift [hunk 3]
```diff
@@ -220,6 +220,46 @@ public class BraveRewards: PreferencesObserver {
...
+
+  /// Notifies Brave Ads that a tab did change
```
Comment (tmancey) [inline ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift:225]: /// Notifies Brave Ads that the given tab did change

File: ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift [hunk 4]
```diff
@@ -220,6 +220,46 @@ public class BraveRewards: PreferencesObserver {
...
+    ads.notifyTabDidChange(
+      tabId,
+      redirectChain: tab.redirectChain,
+      isNewNavigation: tab.rewardsReportingState.isNewNavigation,
+      isRestoring: tab.rewardsReportingState.wasRestored,
+      isSelected: isSelected
+    )
+  }
+
+  /// Notifies Brave Ads that a tab did load
```
Comment (tmancey) [inline ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift:247]: /// Notifies Brave Ads that the given tab did load

File: ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift [hunk 5]
```diff
@@ -220,6 +220,46 @@ public class BraveRewards: PreferencesObserver {
...
+    tab: Tab,
+    isSelected: Bool
+  ) {
+    guard !tab.redirectChain.isEmpty, !tab.isPrivate, ads.isServiceRunning() else {
+      // Don't notify `DidChange` for tabs that haven't finished loading, private tabs,
+      // or when the ads service is not running.
+      return
+    }
+
+    let tabId = Int(tab.rewardsId)
```
Comment (tmancey) [inline ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift:236]: Do we need to assign tabId, can we not just use `Int(tab.rewardsId)`?

File: ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift [hunk 6]
```diff
@@ -220,6 +220,46 @@ public class BraveRewards: PreferencesObserver {
...
+
+  /// Notifies Brave Ads that a tab did load
+  func maybeNotifyTabDidLoad(tab: Tab) {
+    guard !tab.redirectChain.isEmpty, !tab.isPrivate, ads.isServiceRunning() else {
+      // Don't notify `DidLoad` for tabs that haven't finished loading, private tabs,
+      // or when the ads service is not running.
+      return
+    }
+
+    let tabId = Int(tab.rewardsId)
```
Comment (tmancey) [inline ios/brave-ios/Sources/Brave/Frontend/Rewards/BraveRewards.swift:255]: Do we need to assign tabId, can we not just use `Int(tab.rewardsId)`?

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**